include/boost/url/bnf/parse.hpp

```diff
@@ -77,214 +96,182 @@ parse(
-     prefer setting @ref ec instead.
 
-   @see https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form
+     Internal calls to `parse` overloads for `args` may throw, although they should
+     prefer setting `ec` instead.
+
+   @see
+
+   [Backus–Naur form](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form),
 
-        @ref boost::urls::error
+   @ref boost::urls::error
 
    @return `true` if the range initial chars match the elements successfully,
    `false` otherwise
 
-   @param[in,out] it An iterator to the first element in the range.
-   At the end of this function, @ref it points to one past the last
+   @param it An iterator to the first element in the range.
+   At the end of this function, `it` points to one past the last
    element parsed in the range.
 
-   @param[in] end An iterator to one past the last element in the range
+   @param end An iterator to one past the last element in the range
 
-   @param[out] ec The error code this function should set if it fails
+   @param ec The error code this function should set `it` it fails
 
-   @param[out] args BNF elements to be parsed
+   @param args BNF elements to be parsed
 
 */
+template<
 #ifdef BOOST_URL_DOCS
-template<class... Args>
-bool parse( char const*& it, char const* end, Args&&... args);
+    class... Args
 #else
-template<
     class T0,
     class T1,
-    class... Tn>
+    class... Tn
+#endif
+>
 bool
 parse(
     char const*& it,
     char const* end,
     error_code& ec,
+#ifdef BOOST_URL_DOCS
+    Args&&... args
+#else
     T0&& t0,
     T1&& t1,
-    Tn&&... tn);
+    Tn&&... tn
 #endif
+);
 
-/** @brief Parse a complete string
+/** Parse a complete string as a sequence of BNF elements
 
-   This function parses a complete string @ref s into the Backus–Naur form (BNF)
-   elements [@ref t0, @ref t1, ...]. If the string is not completely consumed,
-   @ref ec is set to @ref error::leftover_input.
+   This function parses a complete string `s` into the Backus–Naur form (BNF)
+   elements [`arg_0`, `arg_1`, ...] in `args`.
 
-   The function returns `true` and resets @ref ec when the input is
-   parsed successfully into the elements [@ref t0, @ref t1, ...].
+   @par Example
 
-   If the input cannot be parsed into the BNF elements, the function returns
-   `false` and sets the error number in @ref ec to the appropriate error
-   number defined in @ref boost::urls::error.
+   @code
+   if (bnf::parse(str, ec, bnf_element_1, bnf_element_2, bnf_element_3)) {
+     std::cout << "String " << str << " parsed successfully\n";
+   }
+   @endcode
 
    @par Exception Safety
-     Calls to `parse` overloads may throw, although they should
-     prefer setting @ref ec instead.
 
-   @see @ref boost::urls::error
+     Internal calls to `parse` overloads may throw, although they should
+     prefer setting `ec` instead.
 
-   @return `true` if the string matches the BNF elements successfully and all
-   chars are consumed, `false` otherwise
+   @see
```

Comment:
Ordering: I put `@see` at the end, and it only lists refs (`@ref`) never hyperlinks. If you want to include hyperlinks, which is fine, do it this way
```
@par References
    @li <href="https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form">Backus–Naur form (Wikipedia)</a>
```

Reply:
Put References just before See Also. Alternatively you could use the paragraph title `@par Specification`.

Reply:
OK. I must have made a mistake using the template I found in `javadoc.hpp`. It seems like the file has changed since then. I'll recreate the template.

Reply:
javadoc.hpp will probably always be hopelessly out of date. It means well, but...

Reply:
No problem. But please don't give up on it. It's still the best reference to get started because everyone has a different style.  

I use the bulky `///` comments in my libs. You would probably hate that too. :laughing:

Reply:
`///` makes things harder to edit since you have to fiddle with the comments on every line

Reply:
Yep. `///` is just an example of why `javadoc.hpp` is important.

> /// makes things harder to edit since you have to fiddle with the comments on every line

Not in my case. The IDE and clang-format do everything for me and it's impossible for contributors to ever get it wrong. 

On the other hand, the IDE always gets `/** */` and no `///` wrong for me and I always have to readjust the indentation line by line to make the text start where it should. `///` and `*` on each line, like ASIO does, works a little better though.

I also find `///` (or even `///` + `*`) more aesthetically useful to identify documentation blocks when reading header files, like in ASIO. 

None of that is an argument, of course. Just an example related to `javadoc.hpp`.